perf(forms): optimize internal method _anyControls in FormGroup

lazarljubenovic opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

We loop over **all** items in search of **at least one** item that satisfies a condition.

Issue Number: N/A


## What is the new behavior?

We stop looping as soon as we find one item.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

I'd suggest changing the name of the method to `_someControls`, since it's clearer what it does in JavaScript: it's more in line with existing `Array#some`, which is also internally used in `FormArray`: https://github.com/angular/angular/blob/df5924abd0c4efe87e2fd189739e565346276b91/packages/forms/src/model.ts#L1987-L1990

alfaproject commented on packages/forms/src/model.ts:
```diff
@@ -1555,11 +1555,12 @@ export class FormGroup extends AbstractControl {
     this._forEachChild((control: AbstractControl, name: string) => {
-      res = res || (this.contains(name) && condition(control));
+      if (this.contains(name) && condition(control)) {
+        return true
```
This is the return of the closure and not `_anyControls`
What you want to do is keep what was there and just add the extra condition:
```ts
if (res) {
  return;
}
```

lazarljubenovic replied:
Ouch, didn't pay attention. Thanks for the review.

JoostK replied:
Actually I don't think you can get this to work as you intended, as `Array.prototype.forEach` as used in `_forEachChild` does not support breaking from the loop early ([see the note in yellow on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)).

lazarljubenovic replied:
Indeed, it just exits the inner lambda and does nothing. I guess we I could just transform it to use a regular `for-of` loop in this case, since it's already done a few lines below anyway, for the same reason:

https://github.com/angular/angular/blob/a198c40787e5bf2330d55953cb777374cee3fa91/packages/forms/src/model.ts#L1585-L1593

lazarljubenovic committed:
perf(forms): optimize internal method _anyControls in FormGroup

The method was previously looping through all controls, even after finding at least one that
satisfies the provided condition. This can be a bottleneck with large forms. The new version
of the method returns as soon as a single control which conforms to the condition is found.

AndrewKushnir reviewed (approved):
Thanks for the performance improvement @lazarljubenovic 👍 The change looks good!

Could you please rebase this PR and add a commit message body of at least 100 chars (otherwise the "lint" CI job will trigger an error after rebase)?

Thank you.

AndrewKushnir commented:
[Presubmit](http://test/OCL:315601638:BASE:315601653:1591751977680:82d6c218)

AndrewKushnir commented:
Hi @lazarljubenovic, thanks for updating the commit message. FYI, presubmit was successful, so I'm marking this PR as ready for merge. Thanks for your contribution to Angular!
